File: server/src/main/java/dev/slimevr/autobone/errors/SlideError.java
package dev.slimevr.autobone.errors;

import dev.slimevr.autobone.AutoBoneTrainingStep;
import dev.slimevr.vr.processor.skeleton.HumanSkeleton;
import dev.slimevr.vr.trackers.ComputedTracker;
import dev.slimevr.vr.trackers.TrackerRole;


// The change in position of the ankle over time
public class SlideError implements IAutoBoneError {
	@Override
	public float getStepError(AutoBoneTrainingStep trainingStep) throws AutoBoneException {
		return getSlideError(trainingStep.getSkeleton1(), trainingStep.getSkeleton2());
	}

	public static float getSlideError(HumanSkeleton skeleton1, HumanSkeleton skeleton2) {
		// Calculate and average between both feet
		return (getSlideError(skeleton1, skeleton2, TrackerRole.LEFT_FOOT)
			+ getSlideError(skeleton1, skeleton2, TrackerRole.RIGHT_FOOT)) / 2f;
	}

	public static float getSlideError(
		HumanSkeleton skeleton1,
		HumanSkeleton skeleton2,
		TrackerRole trackerRole
	) {
		// Calculate and average between both feet
		return getSlideError(
			skeleton1.getComputedTracker(trackerRole),
			skeleton2.getComputedTracker(trackerRole)
		);
	}

	public static float getSlideError(ComputedTracker tracker1, ComputedTracker tracker2) {
		// Return the midpoint distance
		return tracker1.position.distance(tracker2.position) / 2f;
	}
}

File: server/core/src/main/java/dev/slimevr/autobone/errors/SlideError.kt
package dev.slimevr.autobone.errors

import dev.slimevr.autobone.AutoBoneStep
import dev.slimevr.tracking.processor.Bone
import dev.slimevr.tracking.processor.BoneType
import dev.slimevr.tracking.processor.skeleton.HumanSkeleton

// The change in position of the ankle over time
class SlideError : IAutoBoneError {
	@Throws(AutoBoneException::class)
	override fun getStepError(trainingStep: AutoBoneStep): Float = getSlideError(
		trainingStep.skeleton1.skeleton,
		trainingStep.skeleton2.skeleton,
	)

	companion object {
		fun getSlideError(skeleton1: HumanSkeleton, skeleton2: HumanSkeleton): Float {
			// Calculate and average between both feet
			return (
				getSlideError(skeleton1, skeleton2, BoneType.LEFT_LOWER_LEG) +
					getSlideError(skeleton1, skeleton2, BoneType.RIGHT_LOWER_LEG)
				) / 2f
		}

		fun getSlideError(
			skeleton1: HumanSkeleton,
			skeleton2: HumanSkeleton,
			bone: BoneType,
		): Float {
			// Calculate and average between both feet
			return getSlideError(
				skeleton1.getBone(bone),
				skeleton2.getBone(bone),
			)
		}

		fun getSlideError(bone1: Bone, bone2: Bone): Float {
			// Return the midpoint distance
			return (bone2.getTailPosition() - bone1.getTailPosition()).len() / 2f
		}
	}
}
